pkg/envoy/registry/registry.go

```diff
@@ -10,43 +10,64 @@ func NewProxyRegistry(mapper ProxyServiceMapper, msgBroker *messaging.Broker) *P
 	return &ProxyRegistry{
 		ProxyServiceMapper: mapper,
 		msgBroker:          msgBroker,
+		connectedProxies:   make(map[string]*envoy.Proxy),
 	}
 }
 
 // RegisterProxy registers a newly connected proxy.
 func (pr *ProxyRegistry) RegisterProxy(proxy *envoy.Proxy) {
-	pr.connectedProxies.Store(proxy.UUID.String(), proxy)
+	pr.countLock.Lock()
+	defer pr.countLock.Unlock()
+
+	uuid := proxy.UUID.String()
+	if pr.GetConnectedProxy(uuid) != nil {
+		log.Debug().Str("proxy", proxy.String()).Msgf("Proxy %s already registered", proxy.String())
+		return
+	}
+
+	pr.connectedProxies[uuid] = proxy
+	pr.proxiesCount++
 	log.Debug().Str("proxy", proxy.String()).Msg("Registered new proxy")
 }
 
 // GetConnectedProxy loads a connected proxy from the registry.
 func (pr *ProxyRegistry) GetConnectedProxy(uuid string) *envoy.Proxy {
-	p, ok := pr.connectedProxies.Load(uuid)
+	p, ok := pr.connectedProxies[uuid]
 	if !ok {
 		return nil
 	}
-	return p.(*envoy.Proxy)
+	return p
 }
 
 // UnregisterProxy unregisters the given proxy from the catalog.
-func (pr *ProxyRegistry) UnregisterProxy(p *envoy.Proxy) {
-	pr.connectedProxies.Delete(p.UUID.String())
-	log.Debug().Msgf("Unregistered proxy %s", p.String())
+func (pr *ProxyRegistry) UnregisterProxy(proxy *envoy.Proxy) {
+	pr.countLock.Lock()
+	defer pr.countLock.Unlock()
+
+	uuid := proxy.UUID.String()
+	if pr.GetConnectedProxy(uuid) == nil {
+		log.Debug().Str("proxy", proxy.String()).Msgf("Proxy %s does not exist", proxy.String())
+		return
+	}
+
+	delete(pr.connectedProxies, uuid)
+	pr.proxiesCount--
+	log.Debug().Msgf("Unregistered proxy %s", proxy.String())
 }
 
 // GetConnectedProxyCount counts the number of connected proxies
-// TODO(steeling): switch to a regular map with mutex so we can get the count without iterating the entire list.
 func (pr *ProxyRegistry) GetConnectedProxyCount() int {
-	return len(pr.ListConnectedProxies())
+	pr.countLock.RLock()
```

Comment:
there's actually 2 locks associated with a read lock. Read locks are great for when you do a lot of reads, and your ops (read/write), might have some longer blocking function. In this case, since it's just setting/deleting/reading values it should be faster to use a standard lock